src/test/java/com/google/devtools/build/lib/actions/ActionCacheCheckerTest.java

```diff
@@ -385,12 +442,475 @@ public void testDeletedConstantMetadataOutputCausesReexecution() throws Exceptio
+
+    // Not cached.
+    runAction(action);
+
+    assertThat(output.getPath().exists()).isFalse();
+    ActionCache.Entry entry = cache.get(output.getExecPathString());
+    assertThat(entry).isNotNull();
+    assertThat(entry.getOutputTree(output)).isNull();
+    assertStatistics(0, new MissDetailsBuilder().set(MissReason.NOT_CACHED, 1).build());
+  }
+
+  @Test
+  public void saveOutputMetadata_treeMetadata_remoteFileMetadataLoaded() throws Exception {
+    cacheChecker = createActionCacheChecker(/* storeOutputMetadata= */ true);
+    SpecialArtifact output =
+        createTreeArtifactWithGeneratingAction(artifactRoot, PathFragment.create("bin/dummy"));
+    ImmutableMap<String, FileArtifactValue> children =
+        ImmutableMap.of(
+            "file1", createRemoteFileMetadata("content1"),
+            "file2", createRemoteFileMetadata("content2"));
+    Action action =
+        new InjectOutputTreeMetadataAction(
+            output, createTreeMetadata(output, children, Optional.empty()));
+    MetadataHandler metadataHandler = new FakeMetadataHandler(derivedPathPrefix);
+
+    runAction(action);
+    Token token =
+        cacheChecker.getTokenIfNeedToExecute(
+            action, null, ImmutableMap.of(), null, metadataHandler, null, ImmutableMap.of());
+
+    TreeArtifactValue expectedMetadata = createTreeMetadata(output, children, Optional.empty());
+    assertThat(token).isNull();
+    assertThat(output.getPath().exists()).isFalse();
+    ActionCache.Entry entry = cache.get(output.getExecPathString());
+    assertThat(entry).isNotNull();
+    assertThat(entry.getOutputTree(output))
+        .isEqualTo(SerializableTreeArtifactValue.createSerializable(expectedMetadata).get());
+    assertThat(metadataHandler.getTreeArtifactValue(output)).isEqualTo(expectedMetadata);
+  }
+
+  @Test
+  public void saveOutputMetadata_treeMetadata_localFileMetadataLoaded() throws Exception {
+    cacheChecker = createActionCacheChecker(/* storeOutputMetadata= */ true);
+    SpecialArtifact output =
+        createTreeArtifactWithGeneratingAction(artifactRoot, PathFragment.create("bin/dummy"));
+    ImmutableMap<String, FileArtifactValue> children1 =
+        ImmutableMap.of(
+            "file1", createRemoteFileMetadata("content1"),
+            "file2", createRemoteFileMetadata("content2"));
+    ImmutableMap<String, FileArtifactValue> children2 =
+        ImmutableMap.of(
+            "file1", createRemoteFileMetadata("content1"),
+            "file2", createRemoteFileMetadata("modified"));
+    Action action =
+        new InjectOutputTreeMetadataAction(
+            output,
+            createTreeMetadata(output, children1, Optional.empty()),
+            createTreeMetadata(output, children2, Optional.empty()));
+    MetadataHandler metadataHandler = new FakeMetadataHandler(derivedPathPrefix);
+
+    // Not cached
+    runAction(action);
+    writeIsoLatin1(output.getPath().getRelative("file2"), "modified");
```

Comment:
Given we create that file, does it matter what children2 contains? I think you could make a stronger test case if those didn't happen to be the same as the state on disk.

Reply:
Changed.